Source/OrchBase/Placement/Algorithms/SimulatedAnnealing.cpp

```diff
@@ -0,0 +1,436 @@
+                        }
+                    }
+                }
+
+                /* And likewise, "unfree" the new core pair for each other
+                 * device type. Handily, set.erase doesn't moan if no element
+                 * matches. */
+                firstCore = selectedThread->parent;
+                secondCore = firstCore->pair;
+
+                for (devtypIt = validCoresForDeviceType.begin();
+                     devtypIt != validCoresForDeviceType.end();
+                     devtypIt++)
+                {
+                    if (devtypIt->first != selectedDevice->pP_devtyp)
+                    {
+                        devtypIt->second.erase(firstCore);
+                        devtypIt->second.erase(secondCore);
+                    }
+                }
+            }
+        }
+
+        /* Transform for swap operation. */
+        else
+        {
+            // <!>
+        }
+
+        iteration++;
+    }
+
+    /* Redistribute devices in cores to evenly load threads. */
+    placer->redistribute_devices_in_task(task);
+
+    /* Write our result structure, and leave. We don't mind that there is a
+     * small difference between the many millions of fitness deltas and the
+     * actual fitness - the difference is relatively small, and is due to
+     * rounding. */
+    float finalFitness = placer->compute_fitness(task);
+    time = placer->timestamp();
+    result.endTime = time;
+    placer->populate_result_structures(&result, task, finalFitness);
+    fprintf(log, "[I] Final fitness from deltas: %f, Iteration count: %d, "
+            "actual final fitness: %f\n",
+            fitness, iteration, finalFitness);
+    fprintf(log, "[I] Placement complete at %s.\n", time.c_str());
+    fclose(log);
+    fclose(data);
+    return fitness;
+}
+
+/* Returns true if the termination condition holds, and false otherwise. It's
+ * pretty simple - for now we're just counting iterations - but this probably
+ * will change as we become more curious. */
+bool SimulatedAnnealing::is_finished(){return iteration >= ITERATION_MAX;}
+
+/* Performs the selection operation for simulated annealing. Arguments
+ *
+ * - task: Holds the graph from which devices are selected.
+ *
+ * - device: Pointer to a device pointer to select for the atomic
+ *   transformation.
+ *
+ * - thread: Pointer to a thread pointer to either move the aforementioned
+ *   device to, or that holds a device to swap with.
+ *
+ * - swapDevice: If the operation is a swap operation, points to the device to
+ *   swap with. If not, is set to PNULL. */
+void SimulatedAnnealing::select(P_task* task, P_device** device,
+                                P_thread** thread, P_device** swapDevice)
+{
+    /* Sanity. */
+    *device = PNULL;
+    *thread = PNULL;
+    *swapDevice = PNULL;
+
+    /* Choose a non-supervisor device. Note that this will loop infinitely if
+     * there are only supervisor devices in this task (but I'm assuming nobody
```

Comment:
Missing a close bracket in the comment, and you know what they say about assumptions...

Reply:
I've fixed this comment, and I've changed simulated annealing to not fall over in this case.